File: processor/src/main/kotlin/io/github/natanfudge/rpc4k/runtime/api/format/JsonFormat.kt
package io.github.natanfudge.rpc4k.runtime.api.format

import kotlinx.serialization.DeserializationStrategy
import kotlinx.serialization.SerializationStrategy
import kotlinx.serialization.json.Json

object JsonFormat : SerializationFormat {
    private val json = Json
    private val charset = Charsets.UTF_8
    override fun <T> encode(serializer: SerializationStrategy<T>, value: T): ByteArray {
        return json.encodeToString(serializer, value).toByteArray(charset)
    }

    override fun <T> decode(serializer: DeserializationStrategy<T>, raw: ByteArray): T {
        return json.decodeFromString(serializer, raw.toString(charset))
    }
}

File: typescript/src/runtime/api/components/JsonFormat.ts
import {SerializationFormat} from "../SerializationFormat";

export const JsonFormat: SerializationFormat = {
    decode<T>(raw: Uint8Array): T {
        return JSON.parse(new TextDecoder().decode(raw)) as T
    },
    encode(value: unknown): Uint8Array {
        return new TextEncoder().encode(JSON.stringify(value))
    }
}